app/src/editor/shared/components/modules/Chart.jsx

```diff
@@ -5,78 +5,105 @@ import useIsMounted from '$shared/hooks/useIsMounted'
+    const id = seriesId(idx)
+
+    const datapoint = [x, y]
+
+    seriesCollection[id] = seriesCollection[id] || {
+        data: [],
+        id,
+        idx,
+        type: 's',
+        ready: false,
+    }
+
+    seriesCollection[id].data.push(datapoint)
+
+    return {
+        series: seriesCollection[id],
+        datapoint,
+    }
+}
+
+const redraw = throttle((chart) => {
+    if (chart) {
+        chart.redraw()
+    }
+}, 250)
 
 const ChartModule2 = (props) => {
     const { isActive, module } = props
 
-    const subscriptionRef = useRef(null)
+    const seriesCollectionRef = useRef({})
 
-    const queuedDatapointsRef = useRef([])
-
-    const [series, setSeries] = useState({})
-
-    const [seriesData, setSeriesData] = useState({})
-
-    const onSeries = useCallback((payload) => {
-        const id = `series-${payload.idx}`
-
-        setSeries((series) => ({
-            ...series,
-            [id]: {
-                ...(series[id] || {}),
-                ...payload,
-                lineWidth: 1,
-                marker: {
-                    lineColor: undefined,
-                    symbol: 'circle',
-                },
-                showInNavigator: true,
-                id,
-                xAxis: 0,
-                yAxis: 0,
-            },
-        }))
-    }, [setSeries])
+    const subscriptionRef = useRef(null)
 
     const isMounted = useIsMounted()
 
-    const flushDatapointsRef = useRef(throttle(() => {
-        if (!isMounted()) {
-            return
-        }
+    const [chart, setChart] = useState(null)
 
-        const queued = queuedDatapointsRef.current || []
-        queuedDatapointsRef.current = []
-
-        setSeriesData((seriesData) => queued.reduce((memo, { s, x, y }) => ({
-            ...memo,
-            [s]: [
-                ...(memo[s] || []),
-                [x, y],
-            ],
-        }), seriesData))
-    }, 250))
-
-    const onDatapoint = useCallback((payload) => {
-        queuedDatapointsRef.current.push(payload)
-        flushDatapointsRef.current()
-    }, [flushDatapointsRef, queuedDatapointsRef])
-
-    const onMessage = useCallback((payload) => {
-        switch (payload.type) {
+    const onMessage = useCallback(({ type, ...payload }) => {
+        let series
+
+        let datapoint
+
+        switch (type) {
             case 'p':
-                onDatapoint(payload)
+                ({ series, datapoint } = addSeriesDatapoint(seriesCollectionRef.current, payload))
                 break
             case 's':
-                onSeries(payload)
+                ({ series } = updateSeriesCollection(seriesCollectionRef.current, payload))
                 break
             default:
-                // noop
         }
-    }, [onDatapoint, onSeries])
 
-    const onSeriesRef = useRef()
-    onSeriesRef.current = onSeries
+        if (!chart || (type !== 'p' && type !== 's')) {
+            return
```

Comment:
maybe add a comment as to when/how this can occur and why no update or if it's just defensive

Reply:
Updated the condition for better readability. No chart or no series? Bail.